codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/EndpointHostPrefixMiddleware.java

```diff
@@ -0,0 +1,204 @@
+                writer.addUseImports(SmithyGoDependency.STRINGS);
+
+                w.write("var prefix strings.Builder");
+                StructureShape input = ProtocolUtils.expectInput(model, operation);
+                for (SmithyPattern.Segment segment : pattern.getSegments()) {
+                    if (!segment.isLabel()) {
+                        w.write("prefix.WriteString($S)", segment.toString());
+                    } else {
+                        MemberShape member = input.getMember(segment.getContent()).get();
+                        String memberName = symbolProvider.toMemberName(member);
+                        String memberReference = "input." + memberName;
+
+                        // Theoretically this should never be nil or empty by this point unless validation has
+                        // been disabled.
+                        w.write("if $L == nil || len(*$L) == 0 {", memberReference).indent();
+                        w.write("return out, metadata, &smithy.SerializationError{Err: "
+                                        + "fmt.Errorf(\"$L forms part of the endpoint host and so may not be nil\")}",
+                                memberName);
+                        w.dedent().write("} else if !smithyhttp.ValidHostLabel(*$L) {", memberReference).indent();
+                        w.write("return out, metadata, &smithy.SerializationError{Err: "
+                                + "fmt.Errorf(\"$L forms part of the endpoint host and so must match "
+                                + "\\\"[a-zA-Z0-9-]{1,63}\\\""
+                                + ", but was \\\"%s\\\"\", *$L)}", memberName, memberReference);
+                        w.dedent().openBlock("} else {", "}", () -> {
+                            w.write("prefix.WriteString(*$L)", memberReference);
+                        });
+                    }
+                }
+
+                w.write("request.HostPrefix = prefix.String()");
+            }
+            w.write("");
+
+            w.write("return next.$L(ctx, in)", generator.getHandleMethodName());
+        });
+    }
+
+    /**
+     * Gets a list of the operations decorated with the EndpointTrait.
+     *
+     * @param model   Model used for generation.
+     * @param service Service for getting list of operations.
+     * @return list of operations decorated with the EndpointTrait.
+     */
+    public static List<OperationShape> getOperationsWithEndpointPrefix(Model model, ServiceShape service) {
+        List<OperationShape> operations = new ArrayList<>();
+        service.getAllOperations().stream().forEach((operationId) -> {
+            OperationShape operation = model.expectShape(operationId).asOperationShape().get();
+            if (!operation.hasTrait(EndpointTrait.ID)) {
+                return;
+            }
+
+            operations.add(operation);
+        });
+        return operations;
+    }
```

Comment:
Instead of `getOperationsWithEndpointPrefix` method, returning a list of operations, suggest changing this method to `isOperationWithEndpointPrefix` which takes in an operation and return a boolean indicating if an operation has endpoint trait. 

Right now you call this method twice, you iterate all operations twice and then again iterate on returned operations list.

Reply:
Posted another alternative above, where you can reuse the previously computed operation list

Reply:
The intention of this helper is a quick get list of operations that are match the filter and iterate over the result. Similar to how the knowledge indexes, and other collectors behave. This also would apply to other modules like the SDK's codegen that may use this helper to iterate through the operations with this trait in order to generate unit tests for operations using this trait.

Will investigate the is filter vs the collector pattern